streaming_form_data/_parser.pyx

```diff
@@ -247,62 +252,134 @@ cdef class _Parser:
+                        return 100
+                    match_start = idx + 1 - self.ender_length
 
-                        if chunk[_idx - 1] == Constants.LF and \
-                                chunk[_idx - 2] == Constants.CR:
-                            self.on_body(chunk[buffer_start: _idx - 2])
-                        else:
-                            self.on_body(chunk[buffer_start: _idx])
+                    if match_start >= buffer_start:
+                         self.on_body(chunk[buffer_start: match_start])
+                    else:
+                        return 110
 
-                        buffer_start = buffer_end = idx + 1
+                    buffer_start = idx + 1
 
                     self.unset_active_part()
                     self.ender_finder.reset()
+
                 else:
-                    if self.ender_finder.inactive() and \
-                            self.delimiter_finder.inactive() and \
-                            buffer_end - buffer_start > Constants.MaxBufferSize:
-                        _idx = buffer_end - 1
+                    # The following block is for great speed optimization
+                    # The idea is to skip all data not containing
+                    # delimiter starting sequence '\r\n--' when
+                    # we are not already in the middle of potential delimiter
 
-                        self.on_body(chunk[buffer_start: _idx])
+                    if self.delimiter_finder.inactive():
+                        skip_count = self.RewindFastForward(chunk_ptr, idx + 1, chunk_len-1)
+                        idx += skip_count
 
-                        buffer_start = idx
             elif self.state == ParserState.PS_END:
                 return 0
             else:
-                return 1
+                return 120
+
+            idx += 1
+
+        if idx != chunk_len:
+            return 140
+        if buffer_start > chunk_len:
+            return 150
 
-        if buffer_end - buffer_start > 0:
-            self._leftover_buffer = chunk[buffer_start: buffer_end]
+        if self.state == ParserState.PS_READING_BODY:
+            matched_length = max(self.delimiter_finder.matched_length(),
+                                 self.ender_finder.matched_length())
+            match_start = idx - matched_length
+            if match_start >= buffer_start + Constants.MinFileBodyChunkSize:
+                self.on_body(chunk[buffer_start: match_start])
+                buffer_start = match_start
+
+        if idx - buffer_start > 0:
+            self._leftover_buffer = chunk[buffer_start: idx]
 
         return 0
+
+    # RewindFastForward is searching for "\r\n--" sequence in provided buffer.
+    # It returns number of chars which can be skipped before delimiter starts (including potential 4-byte match).
+    # It may also update Finder object state.
+    cdef Index RewindFastForward(self, const Byte *chunk_ptr, Index posFirst, Index posLast):
```

Comment:
Please rename `posFirst` to `pos_first`, and `posLast` to `pos_last` for consistency with the rest of the codebase.

Reply:
sure. My mistake

Reply:
No worries!